lemmy_apub/src/objects/mod.rs

```diff
@@ -127,3 +169,39 @@ pub(in crate::objects) fn check_is_markdown(mime: Option<&Mime>) -> Result<(), L
     Ok(())
   }
 }
+
+/// Converts an ActivityPub object (eg `Note`) to a database object (eg `Comment`). If an object
+/// with the same ActivityPub ID already exists in the database, it is returned directly. Otherwise
+/// the apub object is parsed, inserted and returned.
+pub(in crate::objects) async fn get_object_from_apub<From, Kind, To, ToForm>(
+  from: &From,
+  context: &LemmyContext,
+  expected_domain: Option<Url>,
+  request_counter: &mut i32,
+) -> Result<To, LemmyError>
+where
+  From: BaseExt<Kind>,
+  To: ApubObject + Crud<ToForm> + Send + 'static,
+  ToForm: FromApubToForm<From> + Send + 'static,
+{
+  let object_id = from.id_unchecked().context(location_info!())?.to_owned();
+  let object_in_database = blocking(context.pool(), move |conn| {
+    To::read_from_apub_id(conn, object_id.as_str())
+  })
+  .await?;
+
+  // if we already have the object in our database, return that directly
+  if let Ok(to) = object_in_database {
+    Ok(to)
+  }
+  // if we dont have it, parse from apub and insert into database
+  // TODO: this is insecure, a `Like/Post` could result in a non-existent post from a different
+  //       instance being inserted into our database. we should request the object over http in that
+  //       case. this might happen exactly in the case where expected_domain = None, but i'm not sure.
```

Comment:
This is the problem I described in my [last comment](https://github.com/LemmyNet/lemmy/pull/1297#issuecomment-737219046). I think to avoid it, we have to change the code below so that it fetches the object over HTTP, specifically in the case where `expected_domain = None`. But I am not sure.